### DISCUSSION TITLE
Fprime for RTEMS

### DISCUSSION THREAD

**Discussion by polaraf** [nasa/fprime#1082] [Q&A]
frpime already ported to rtems  ?i don't find any rtems references in the opensource codebase

**Comment by Joshua-Anderson** [comment]
I believe some F' demos have been ported to RTEMS is the past. Since RTEMS is posix compliant it should be possible to use the posix implementation of tasks and the filesystem and run F' on RTEMS with minimal modifications.

**Reply by ThibFrgsGmz** [reply]
I sense an appetite to implement RTEMS in the community. Me too because traditionally on our products we use RTEMS.

Why is the JPL implementation not included in the open source project? Too difficult to maintain? Is there too much intellectual property?

Would you like a collaborator to carry out that implementation and incorporate it into the project? Is the previous implementation made at your place going to be released as open source one day?

We could take inspiration from @nmullane's work on nasa/cFS to check for non-regressions on the RTEMS support. For the GitHub Actions CI, he implemented the build and unit tests of cFS in RTEMS 4.11/5 in QEMU.

**Comment by Joshua-Anderson** [comment]
> Why is the JPL implementation not included in the open source project?

It wasn't open sourced because it was just a proof-of-concept prototype. I think a lot of it actually made it into old releases of fprime but since it wasn't really being used it bit-rotted and slowly disappeared as parts of fprime were refactored over the years. 

I spent a partial day last year trying to get F' compiling on RTEMS and got pretty close, so I suspect a RTEMS port would be pretty straightforward, at least using the POSIX apis.

> Would you like a collaborator to carry out that implementation and incorporate it into the project?

A RTEMS port would be awesome! We'd be happy to support any interested contributor. There's certainly interest in RTEMs support, there's just a lot of higher priority items on the JPL F' roadmap.

> We could take inspiration from @nmullane's work on nasa/cFS to check for non-regressions on the RTEMS support.

I agree, if you don't regression test your target platforms you're nearly guaranteed to break them. This is a big problem with F' VxWorks support since providing VxWorks in open-source CI gets messy licensing wise. I want to set up github actions to build against actual microcontrollers to test baremetal F'.

**Reply by ThibFrgsGmz** [reply]
I'd be interested in helping with the implementation of this! However, I'm currently working on a scientific publication and could only begin during the winter vacations.

Is it possible to share PoCs or code snippets from back then with the community to determine which parts need to be modified?

I have an RPI 4B, but I don't think I'll be able to install RTEMS on it.

I found a 2013 [tutorial](http://alanstechnotes.blogspot.com/2013/03/rtems-on-raspberry-pi.html) from @acudmore who had set up an RTEMS development environment for RPI back in the day, but it was with an older version of RPI. Furthermore, the RTEMS BSP for RPI 4 is not currently supported, according to the [RTEMS documentation](https://docs.rtems.org/branches/master/user/bsps/arm/raspberrypi.html).

I probably have a way to find the RPI I had while in college but that was the first model (~2012). Despite the fact that the RTEMS documentation states that their BSP begins with RPI 2, Alan must have tested with RPI 1 because the tutorial is from 2013 and RPI 2 from 2015.

**Reply by Joshua-Anderson** [reply]
That would be awesome! If I remember correctly RTEMS also had a few well supported QEMU simulated targets. That might be a good place to get started because it makes it easier for other contributors to test RTEMS without needing matching hardware

**Reply by ThibFrgsGmz** [reply]
Yes, you are entirely correct. I will attempt to program this implementation using GitPod and the same rtems/QEMU container as cFS.
For the time being, however, there is an [issue](https://github.com/nmullane/qemu_rtems_docker/issues/1) with the container when I instantiate it with GitPod. To be continued...
